index.d.ts

```diff
@@ -52,15 +43,18 @@ export interface AnyAction extends Action {
  *
  * *Do not put API calls into reducers.*
  *
- * @template S State object type.
+ * @template S The type of state consumed and produced by this reducer.
+ * @template A The type of actions the reducer can potentially respond to.
  */
-export type Reducer<S> = (state: S, action: AnyAction) => S;
+export type Reducer<S = {}, A extends Action = Action> = (state: S, action: A) => S;
```

Comment:
> Is there a particular reason for `S` to default to `{}`?

Not really, I was just on a defaulting rampage.

> Also, I don't think that it is correct to constraint the type of action. A reducer must be able to accept every possible action, which is any object with type property. It should just ignore all actions that don't affect its state.

It's perfectly valid to constrain the type of actions that a reducer handles. It must be able to accept any action in your app, but not any arbitrary object with a `type` field. This is critical to making the library nice to use in TypeScript because it allows the type system to automatically refine the types for you as you learn information by inspecting the `type` field. For example:

```ts
// The universe of all actions used in your app
export type AnyAction =
  | SetGreeting
  | IncrementCounter
  | SetCompletionStatus

export interface SetGreeting {
  type: 'SET_GREETING'
  greeting: string
}

export interface IncrementCounter {
  type: 'INCREMENT_COUNTER'
  counterIndex: number
}

export interface SetCompletionStatus {
  type: 'SET_COMPLETION_STATUS'
  completed: boolean
}

const reducer: Reducer<State, AnyAction> = (state = initialState, action) => {
  switch (action.type) {
    case: 'SET_GREETING':
      return { ...state, greeting: action.greeting }
    case: 'INCREMENT_COUNTER':
      return { ...state, counters: state.counters[action.counterIndex] + 1 }
    case: 'SET_COMPLETION_STATUS':
      return { ...state, completed: action.completed }
  }
}
```

What's super cool about this is
1. It's completely type safe. Inside each `case`, the type of `action` has been refined such that it knows what fields should be available on it and what types they have. If I tried to access `action.counterIndex` in a branch where I hadn't checked that `action.type === 'INCREMENT_COUNTER'` it would be a type error.
1. We get exhaustiveness checking, so that no `default` case is required because the type system has proven that you've handled all possible action types.
1. You can perform an app-wide refactoring on the string literals being used for the action `type` fields and they will also change these switch cases.